File: sharedUI/src/commonMain/kotlin/dev/yveskalume/elevenlabs/feature/LoadingState.kt
package dev.yveskalume.elevenlabs.feature

sealed interface LoadingState<out T> {
    data class Idle<T>(val data: T) : LoadingState<T>
    data object Loading : LoadingState<Nothing>
    data class Error<T>(val message: String) : LoadingState<T>

    val dataOrNull: T?
        get() = (this as? Idle)?.data
}

File: iosApp/iosApp/Shared/State/LoadingState.swift
enum LoadingState<Content> {
    case loading
    case error(message: String)
    case idle(data: Content)

    var dataOrNil: Content? {
        switch self {
        case let .idle(data):
            data
        case .loading, .error:
            nil
        }
    }
}
